<a href="https://colab.research.google.com/github/elmilanos/ProbabilityStatistics/blob/main/Pr%C3%A1ctica_M%C3%B3dulo_12_An%C3%A1lisis_de_Regresi%C3%B3n_M%C3%BAltiple.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 1. Datos de la Practica

In [1]:
import pandas as pd

# Cargar el archivo Excel proporcionado
file_path = '/Practica_Regresion_Múltiple.xlsx'
data = pd.read_excel(file_path)

# Mostrar las primeras filas para analizar los datos
data.head()


Car       Model  Volume  Weight  CO2
0      Toyota        Aygo    1000     790   99
1  Mitsubishi  Space Star    1200    1160   95
2       Skoda      Citigo    1000     929   95
3        Fiat         500     900     865   90
4        Mini      Cooper    1500    1140  105

Para la regresión múltiple, necesitamos identificar la variable dependiente (Y) y las variables independientes (X). En este caso, podemos usar **CO2** como la variable dependiente, y **Volume** y **Weight** como las variables independientes.

In [11]:
import numpy as np

# Recalcular los coeficientes (β) para asegurarme de que los valores estén definidos
X_matrix = X.values
Y_matrix = Y.values

# Calcular X^T X
X_transpose_X = X_matrix.T.dot(X_matrix)

# Calcular (X^T X)^-1
X_transpose_X_inv = np.linalg.inv(X_transpose_X)

# Calcular X^T Y
X_transpose_Y = X_matrix.T.dot(Y_matrix)

# Calcular los coeficientes (β)
beta = X_transpose_X_inv.dot(X_transpose_Y)

# Realizar predicciones usando el modelo con los coeficientes obtenidos
predictions = X_matrix.dot(beta)

# Calcular R² y MSE
r2 = r2_score(Y_matrix, predictions)
mse = mean_squared_error(Y_matrix, predictions)

# Imprimir los coeficientes y métricas de forma adecuada
print(f"Coeficientes (β): Intercepto = {beta[0]:.4f}, Volumen (β₁) = {beta[1]:.4f}, Peso (β₂) = {beta[2]:.4f}")
print(f"R² (Coeficiente de determinación): {r2:.4f}")
print(f"MSE (Error cuadrático medio): {mse:.4f}")


Coeficientes (β): Intercepto = 79.6947, Volumen (β₁) = 0.0078, Peso (β₂) = 0.0076
R² (Coeficiente de determinación): 0.3766
MSE (Error cuadrático medio): 33.6828


Por lo tanto, la ecuación de regresión que describe la relación entre las emisiones de CO2 y las variables independientes (Volumen y Peso) es:
*CO2*=*79.6947*+*0.0078*⋅*Volume*+*0.0076*⋅*Weight*

## 3. Implementación con **scikit-learn**

In [12]:
# Recalcular los coeficientes y realizar las predicciones
import numpy as np
import pandas as pd
from sklearn.metrics import r2_score, mean_squared_error

# Cargar el archivo Excel nuevamente
file_path = '/Practica_Regresion_Múltiple.xlsx'
data = pd.read_excel(file_path)

# Definir la variable dependiente (Y) y las variables independientes (X)
X = data[['Volume', 'Weight']]  # Variables independientes
Y = data['CO2']  # Variable dependiente

# Agregar una columna de 1s para el intercepto (β0)
X = pd.concat([pd.Series([1]*len(X), name='Intercept'), X], axis=1)

# Calcular los coeficientes usando la fórmula β = (X^T X)^-1 X^T Y
X_matrix = X.values
Y_matrix = Y.values

# Calcular X^T X
X_transpose_X = X_matrix.T.dot(X_matrix)

# Calcular (X^T X)^-1
X_transpose_X_inv = np.linalg.inv(X_transpose_X)

# Calcular X^T Y
X_transpose_Y = X_matrix.T.dot(Y_matrix)

# Calcular los coeficientes (β)
beta = X_transpose_X_inv.dot(X_transpose_Y)

# Realizar predicciones usando el modelo con los coeficientes obtenidos
predictions = X_matrix.dot(beta)

# Calcular R² y MSE
r2 = r2_score(Y_matrix, predictions)
mse = mean_squared_error(Y_matrix, predictions)

beta, r2, mse


(array([7.96947193e+01, 7.80525753e-03, 7.55094727e-03]),
 0.3765564043619989,
 33.68279098995154)

Los resultados de la evaluación del modelo son los siguientes:

- **R²** **(Coeficiente de determinación):** 0.3766, lo que indica que el modelo explica aproximadamente el 37.66% de la variabilidad de las emisiones de CO2 en función del volumen y el peso de los coches.

- **MSE (Error cuadrático medio):** 33.6828, lo que mide la diferencia promedio entre las predicciones del modelo y los valores reales de CO2.

Aunque el R² no es muy alto, esto sugiere que el modelo podría mejorar, posiblemente mediante la inclusión de otras variables o un análisis más profundo de las relaciones entre las variables.